src/mlpack/methods/lmnn/lmnn_function_impl.hpp

```diff
@@ -0,0 +1,415 @@
+void LMNNFunction<MetricType>::Shuffle()
+{
+  arma::mat newDataset;
+  arma::Row<size_t> newLabels;
+
+  math::ShuffleData(dataset, labels, newDataset, newLabels);
+
+  math::ClearAlias(dataset);
+  math::ClearAlias(labels);
+
+  dataset = std::move(newDataset);
+  labels = std::move(newLabels);
+
+  // Re-calculate target neighbors as indices changed.
+  Constraints constraint(dataset, labels, k);
+  constraint.TargetNeighbors(targetNeighbors);
+}
+
+inline void Projection(arma::mat& iterate)
+{
+  // Projection.
+  arma::vec eigval;
+  arma::mat eigvec;
+
+  arma::eig_sym(eigval, eigvec, iterate);
+
+  arma::uvec ind = arma::find(eigval > 0);
+  arma::mat diagEigVal = arma::diagmat(eigval);
+  iterate = eigvec.cols(ind) * diagEigVal.submat(ind, ind) *
+      arma::trans(eigvec.cols(ind));
+}
+
+//! Evaluate cost over whole dataset.
+template<typename MetricType>
+double LMNNFunction<MetricType>::Evaluate(const arma::mat& coordinates)
+{
+  // Hack
+  arma::mat& iterate = const_cast<arma::mat&>(coordinates);
+  // Apply projection.
+  if (arma::accu(iterate == initialPoint) != initialPoint.n_elem)
+    Projection(iterate);
+
+  double cost = 0;
+
+  // Apply metric over dataset.
+  transformedDataset = coordinates * dataset;
+
+  // Re-calculate impostors on transformed dataset.
+  Constraints constraint(transformedDataset, labels, k);
+  constraint.Impostors(impostors);
+
+  for (size_t i = 0; i < dataset.n_cols; i++)
+  {
+    for (size_t j = 0; j < k ; j++)
+    {
+      // Calculate cost due to distance between target neighbors & data point.
+      double eval = metric.Evaluate(transformedDataset.col(i),
+                          transformedDataset.col(targetNeighbors(j, i)));
+      cost += (1 - regularization) * eval;
+    }
+
+    for (size_t j = 0; j < k ; j++)
+    {
+      for (size_t l = 0; l < k ; l++)
+      {
+        // Calculate cost due to data point, target neighbors, impostors
+        // triplets.
+        double eval = metric.Evaluate(transformedDataset.col(i),
+                          transformedDataset.col(targetNeighbors(j, i))) -
+                      metric.Evaluate(transformedDataset.col(i),
+                          transformedDataset.col(impostors(l, i)));
+
+        cost += regularization * std::max(0.0, 1 + eval);
+      }
+    }
+  }
+
+  return cost;
+};
+
+//! Calculate cost over batches.
+template<typename MetricType>
+double LMNNFunction<MetricType>::Evaluate(const arma::mat& coordinates,
+                                          const size_t begin,
+                                          const size_t batchSize)
+{
+  // Hack
+  arma::mat& iterate = const_cast<arma::mat&>(coordinates);
+  // Apply projection.
+  if (arma::accu(iterate == initialPoint) != initialPoint.n_elem)
+    Projection(iterate);
+
+  double cost = 0;
+
+  // Apply metric over dataset.
+  transformedDataset = coordinates * dataset;
+
+  Constraints constraint(transformedDataset, labels, k);
+  constraint.Impostors(impostors);
```

Comment:
Note that we only need some of the impostors---specifically those for points with indices between `begin` and `begin + batchSize`.  So you could reduce the size of the KNN query to save some time.

Reply:
This is nice speedup :)